### ISSUE TITLE
.building<VERSION> lock file does not treat multiple variants atomically

### ISSUE THREAD

**Issue by bfloch** [nerdvegas/rez#536]
The filesystem keeps the version if a .building file for the version is present, but a valid package.py is present in this version.

This is probably to respect canceled builds which did not clean up the .building file. Or maybe it is also assumed that a released variant can technically be available in the resolve regardless of other variants still being build.

The package.py is being written after each variant (adding its variant each time, which by the way is another source of problems that I will reveal soon: Parallel variant build might break ...)

That may be intended but for us, with non-mutually exclusive variants it cause issues as with the default orderer a latest version is preferred even though the variant may not perfectly match (yes, python variants again causing trouble).

I am not sure what the best approach is, if this is only us I would write my own filesystem plugin and replace the default one.

Bottom line is: We would love rez-release with multiple variants to be atomic in the eyes on the .building lock file.

**Comment by bfloch** [comment]
Here is my proposal for a backwards compatible fix:

Introduce a setting called building_lock_atomic_variants = False (default)
- In the .building File append the variant index to be build if there is any. For example:
 .building-1.0.0@0
 .building-1.0.0@1
- If the settings is set: 
  - In the check, check also the "number" of variants in the package file as opposed to just the version + existence package file
- If the setting is not set: 
  - Ignore the `@`-part. Will behave as usual.

Why separate files: In order to allow parallel builds of variants. Having one file may introduce race conditions.
Why only the number: Because in my understanding we don't have the real variants up until we have them.

**Comment by maxnbk** [comment]
I agree with this change, but I think the index approach is flawed, as someone who changes variants in packages regularly and rebuilds. I'd recommend a short-hash of the variant-list-as-string or something along those lines. This way, if you comment out some variants and rebuild them, the proper lock file is still associated with the proper variant.

**Comment by bfloch** [comment]
We have a pretty minimal use case as we do never release individual variants (other than concurrently releasing all) and we never change any release package - since this would mess with our timestamp-based environment management system.
But you are right. Probably the mechanism needs more thought.

**Comment by nerdvegas** [comment]
I see the problem, thanks for highlighting it. Nothing to say as yet, I'm going to think on it.

**Comment by bfloch** [comment]
Other then fixing the locking schema/keys I wonder if it wouldn't make sense to abstract the locking in a way that an optional distributed locking server could be used. That would probably make more sense for facilities that distribute their repos. Also it could prove more reliable than filesystem atomicity.

Thoughts?

**Comment by maxnbk** [comment]
Ultimately I want this feature so that I can run variant-builds as renderfarm jobs, so filesystem will be enough for me, but service-based would fulfill my needs as well. (Mostly.)

**Comment by nerdvegas** [comment]
So, there are two separate issues here, it would be good to ensure we aren't talking about both at the same time.

1:
Locking wrt installing new variants: This is already there, and has nothing to do with the `.building` file. Lockfiles are used when updating the package.py, and they act as a mutex. They ensure that during the (read package.py, update with new variant, write new package.py) process, other processes are locked out.

To add to this, a recent rez version has also made package.py writes atomic, which should improve things further (previously, it was in theory possible for a package.py to get read while it's getting rewritten due to variant addition).

2: 
The `.building` file's job is to force rez to check for a present `package.py`, rather than assuming the package is valid just because its version directory is there. IMO it should continue doing only this, rather than being used to force atomicity of multiple variant releases (separation of concerns).

My proposal for supporting atomicity of multiple variant releases: Add a 'pending' mode to the filesystem package repository. This would write installed variants into a separate file (.package.pending.py for eg) that would not affect package/variant ingestion at all. Then we add an "apply_pending" operation to the repo, which takes these pending variants and adds them to the package.py in one atomic step.

It might look something like this:

```
with fs_repo.batch_package_updates():
    install_my_first_variant(fs_repo)
    install_my_second_variant(fs_repo)
```

This would actually be a generic operation to support atomicity of any batched set of package.py updates. This could become useful in other situations also.
